control-plane/p4RuntimeSerializer.cpp

```diff
@@ -639,6 +639,40 @@ getMatchType(cstring matchTypeName) {
     }
 }
 
+static uint32_t
+getNewtypeWidth(const IR::Type* type, TypeMap* typeMap, ReferenceMap* refMap) {
```

Comment:
this function is too complicated. it does not need the referencemap, the typemap should be enough.
it should be called getTypeWidth, and it should report an error for types that do not have a defined width.

Reply:
Please see p4c/testdata/p4_16_samples/p416-type-use3.p4 which Andy provided where `type` is nested with typedef and deeply nested types.  I will think about the rest of your comment.  I can rename the function.

Reply:
typeMap supports getting canonical type from a type.  

refmap is needed because of this portion of the Type_NewType IR.  I need to get the declaration from the path.  From the declaration, I have to recurse for `type` or typedef.
```
type: [189] Type_Name
    path: [188] Path name=PortIdUint_t absolute=0
```

The full IR is included below.

```
[192] Type_Newtype name=PortId_t declid=14
  annotations: [190] Annotations
    [180] Annotation name=p4runtime_translation needsParsing=1
      [177] AnnotationToken token_type=357 text=p4.org/psa/v1/PortId_t constInfo=0
      [178] AnnotationToken token_type=302 text=, constInfo=0
      [179] AnnotationToken token_type=358 text=32 constInfo=0x5558f046c6e0
  type: [189] Type_Name
    path: [188] Path name=PortIdUint_t absolute=0
```

To save one from going to the test file, a portion of the deeply nested type/typedef from Andy is included below.  We need such a recursive function and also the nature of the Type_NewType IR has made the function the way it is.

```
// D is abbreviation for `typedef`
// T is abbrevation for `type`

// Since types and typedefs can be defined in terms of each other, the
// names I use here contain sequences of Ds and Ts to indicate the
// order in which they have been "stacked", e.g. EthDT_t is a `type`
// (the T is last) defined on type of a `typedef` (the D just before
// the T).

typedef bit<48> EthD_t;
type    bit<48> EthT_t;

typedef bit<8>     CustomD_t;
type    bit<8>     CustomT_t;
typedef CustomD_t  CustomDD_t;
type    CustomD_t  CustomDT_t;
typedef CustomT_t  CustomTD_t;
type    CustomT_t  CustomTT_t;
typedef CustomDD_t CustomDDD_t;
type    CustomDD_t CustomDDT_t;
typedef CustomDT_t CustomDTD_t;
type    CustomDT_t CustomDTT_t;
typedef CustomTD_t CustomTDD_t;
type    CustomTD_t CustomTDT_t;
typedef CustomTT_t CustomTTD_t;
type    CustomTT_t CustomTTT_t;
```

Reply:
Note that while I do not consider the test code written above as something I want to run off today and include in some P4_16 program, at least the parts `typedef bit<8> CustomD_t` and `type CustomD_t CustomDT_t` have direct analogous lines in the psa.p4 include file today, as suggested by Mihai.

The rest of the definitions are just me trying stuff out and seeing if anything breaks.

Reply:
It does make sense to use such code to test p4c changes being made.